odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,523 @@
+
+        The abbreviations are "a" for "approximation" and "d" for "detail".
+        The letter position encodes the coefficient type in the respective
+        axis. ``N`` stands for the number of scaling levels.
+    """
+    wbasis = pywt_wbasis(wbasis)
+    coeff_list = []
+
+    wcoeffs = pywt.dwtn(arr, wbasis, pywt_mode)
+    aaa = wcoeffs['aaa']
+    aad = wcoeffs['aad']
+    ada = wcoeffs['ada']
+    add = wcoeffs['add']
+    daa = wcoeffs['daa']
+    dad = wcoeffs['dad']
+    dda = wcoeffs['dda']
+    ddd = wcoeffs['ddd']
+
+    details = (aad, ada, add, daa, dad, dda, ddd)
+    coeff_list.append(details)
+
+    for _ in range(1, nscales):
+        wcoeffs = pywt.dwtn(aaa, wbasis, pywt_mode)
+        aaa = wcoeffs['aaa']
+        aad = wcoeffs['aad']
+        ada = wcoeffs['ada']
+        add = wcoeffs['add']
+        daa = wcoeffs['daa']
+        dad = wcoeffs['dad']
+        dda = wcoeffs['dda']
+        ddd = wcoeffs['ddd']
+        details = (aad, ada, add, daa, dad, dda, ddd)
+        coeff_list.append(details)
+
+    coeff_list.append(aaa)
+    coeff_list.reverse()
+
+    return coeff_list
+
+
+def _waverec3(coeff_list, wbasis, pywt_mode, nscales):
+    """Discrete 3D multiresolution wavelet reconstruction.
+
+    Helper function for the 3D wavelet reconstruction that uses
+    the same type of ordered coefficient list as the 1D and 2D variants
+    of `pywt.waverecn`. This is necessary since the 3D variant takes
+    a dictionary instead of an ordered list.
+
+    Parameters
+    ----------
+    coeff_list : ordered `sequence`
+        Sequence of coefficients organized in the following way:
+
+            ```[aaaN, (aadN, adaN, addN, daaN, dadN, ddaN, dddN), ...,
+            (aad1, ada1, add1, daa1, dad1, dda1, ddd1)]```
+
+        The abbreviations are "a" for "approximation" and "d" for "detail".
+        The letter position encodes the coefficient type in the respective
+        axis. ``N`` stands for the number of scaling levels.
+
+    wbasis :  string or `pywt.Wavelet`
+        Specification of the wavelet to be used in the transform.
+        If a string is given, it is converted to a `pywt.Wavelet`.
+    pywt_mode: string
+        `pywt` style boundary condition.
+    nscales : int
+        Number of scaling levels to be used in the transform.
+        The maximum number of usable scales can be determined
+        by `pywt.dwt_max_level`.
+
+    Returns
+    -------
+    rec : `numpy.ndarray`
+        Wavelet reconstruction from the coefficients.
+    """
+    wbasis = pywt_wbasis(wbasis)
+
+    aaa = coeff_list[0]
+    (aad, ada, add, daa, dad, dda, ddd) = coeff_list[1]
+    coeff_dict = {'aaa': aaa, 'aad': aad, 'ada': ada, 'add': add,
+                  'daa': daa, 'dad': dad, 'dda': dda, 'ddd': ddd}
+    for tpl in coeff_list[2:]:
+        aaa = pywt.idwtn(coeff_dict, wbasis, pywt_mode)
+        (aad, ada, add, daa, dad, dda, ddd) = tpl
+        coeff_dict = {'aaa': aaa, 'aad': aad, 'ada': ada, 'add': add,
```

Comment:
This is repeatedly overwritten, I feel a bit wierd about this code, needs docstrings.